### ISSUE TITLE
Need help to port Gated Conv model to allennlp

### ISSUE THREAD

**Issue by sai-prasanna** [pytorch/fairseq#255]
I am porting the fairseq conv model to allennlp framework. 

I could get the model to work without any dimension errors etc. But when I start training it the loss keeps growing. I tried to train it first on the actual dataset, with most settings same as default ones in fairseq translate, and learning rate less, there was no signs of loss decreasing in a significant manner.
I am unable to overfit even a small set of 200 sentences (from a grammar error correction dataset). 

I am new to coding complex architectures from scratch. Need help to debug the model. 

1. What are the things that can generally go wrong?
2. Will comparing the outputs of each layers between fairseq implementation and mine make sense if I set the seed as same?
3. Any special tricks to get this fconv architecture to work that I could have missed? (I have initialized the parameters in same way in this repository)

Here is the current allennlp code in a gist - https://gist.github.com/sai-prasanna/0572302ea254c7101f7def68868c054d . Fair warning, in the code  `gated_conv_models.py`  is a frankenstien monster made from copy pasting fairseq  code. I will be creating a proper repository with the code in some time.

**Comment by myleott** [comment]
I'm confused, why not just import and wrap the fairseq encoders/decoders directly? `FConvEncoder` and `FConvDecoder` both inherit from nn.Module. That way you don't need to worry if your implementation matches...

Otherwise, it's generally quite painful to exactly match outputs from different frameworks. For example, random seeds must match, but even then you won't get exactly the same random weights if either framework calls the RNG before initializing the model. You should also turn off all dropout for comparisons.

I haven't used allennlp much, but I'd expect you could wrap the fairseq fconv model doing something like this:
```
from fairseq.data import Dictionary
from fairseq.models.fconv import FConvEncoder, FConvDecoder

d = Dictionary()  # TODO write wrapper to convert allennlp dictionary to fairseq one
src_tokens = torch.LongTensor([
    list(map(d.add_symbol, ['hello', 'world', '</s>'])),
    list(map(d.add_symbol, ['<pad>', 'hello', '</s>'])),
])
src_lengths = torch.LongTensor([3, 2])
prev_output_tokens = torch.LongTensor([
    list(map(d.add_symbol, ['</s>', 'hallo', 'welt'])),
    list(map(d.add_symbol, ['</s>', 'hallo', '<pad>'])),
])

convs = ([(512, 3)] * 9 + [(1024, 3)] * 4 + [(2048, 1)] * 2)
encoder = FConvEncoder(d, embed_dim=768, convolutions=convs)
encoder.num_attention_layers = len(convs)
decoder = FConvDecoder(d, embed_dim=768, convolutions=convs)

encoder_out = encoder(src_tokens, src_lengths)
decoder_out = decoder(prev_output_tokens, encoder_out)
lprobs = torch.nn.functional.log_softmax(decoder_out[0], dim=2)
```

**Comment by sai-prasanna** [comment]
@myleott Thanks, I figured out pretty much the same.

I was started the qasi-reimplementation because I figured it would help to have embedding in allennlp's classes. (as they provided more options) This lead to rabbit hole of trying to reimplement everything.

Will get back to you on the wrapped implementation.
